docs(sources/logstash): validated Logstash → OTel Collector → GlassFlow pattern (ETL-1134) (#761)
glassflow/clickhouse-etl: docs/app/sources/logstash/page.mdx

Logstash is a data processing pipeline that ingests, transforms, and forwards logs and events. It is commonly used in the Elastic ecosystem for centralized log management.

## Why direct Logstash → GlassFlow isn't supported today

GlassFlow's OTLP receiver requires the `x-glassflow-pipeline-id` HTTP header on every request. Logstash itself does not have a maintained output plugin that can satisfy this requirement:

> - **No official Elastic OTLP output plugin** exists. Elastic's [own guidance](https://www.elastic.co/observability-labs/blog/logstash-to-otel) is to migrate Logstash pipelines to the OpenTelemetry Collector.

Reviewer: ```suggestion
- **No official Elastic OTLP output plugin** exists. Elastic's [own guidance](https://www.elastic.co/observability-labs/blog/logstash-to-otel) suggests migrating Logstash pipelines to the OpenTelemetry Collector.
```

- The community [`logstash-output-opentelemetry`](https://github.com/paulgrav/logstash-output-opentelemetry) gem hasn't shipped a release since January 2022 and does not support setting custom HTTP headers.

> - The bundled `logstash-output-http` plugin can set custom headers but cannot emit OTLP-protobuf bodies, it has no codec for OTLP encoding.

Reviewer: ```suggestion
- The bundled `logstash-output-http` plugin can set custom headers, but cannot emit OTLP-protobuf bodies, it has no codec for OTLP encoding.
```

# logstash.conf
output {
  tcp {
    host => "otel-collector"
    port => 5044
    codec => line
  }
}
```

### Sink mapping

When you create the GlassFlow pipeline, map **`observed_timestamp`** rather than `timestamp` in the sink mapping. The OTel Collector's `tcplog` receiver does not parse a timestamp from raw Logstash lines, so `LogRecord.Timestamp` is empty; `observed_timestamp` is set by the receiver when the event arrives and is the right value to write to ClickHouse.

```json
"mapping": [
  {"name": "observed_timestamp", "column_name": "timestamp",     "column_type": "DateTime64(9)"},
  {"name": "severity_text",      "column_name": "severity_text", "column_type": "String"},
  {"name": "body",               "column_name": "body",          "column_type": "String"}
]
```

If you want a structured `severity_text` or per-field columns, add an OTel Collector [transform processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor) between `tcplog` and `otlphttp` to extract fields from the raw Logstash payload. The full OTLP logs schema is documented at [OTLP Schema Reference](/sources/otlp/schema-reference#otlp-logs-fields).

> See [Sending data to GlassFlow OTLP](/sources/otlp/sending-data) for full OTel Collector configuration options — including gRPC, TLS, and back-pressure handling.

Reviewer: ```suggestion
See [Sending data to GlassFlow OTLP](/sources/otlp/sending-data) for full OTel Collector configuration options, including gRPC, TLS, and back-pressure handling.
```
